File: version4j/src/main/java/com/only4/algorithm/version4j/leetcode/DiameterOfBinaryTree.java
package com.only4.algorithm.version4j.leetcode;

import com.only4.algorithm.version4j.extra.TreeNode;

public class DiameterOfBinaryTree {
    private int maxDiameter = 0;

    public int diameterOfBinaryTree(TreeNode root) {
        calculateDepth(root);
        return maxDiameter;
    }

    private int calculateDepth(TreeNode node) {
        if (node == null) return 0;

        int leftDepth = calculateDepth(node.left);
        int rightDepth = calculateDepth(node.right);

        // 更新最大直径
        maxDiameter = Math.max(maxDiameter, leftDepth + rightDepth);

        // 返回当前节点的最大深度
        return Math.max(leftDepth, rightDepth) + 1;
    }
}

File: version4k/src/main/kotlin/com/only4/algorithm/version4k/leetcode/diameterOfBinaryTree.kt
package com.only4.algorithm.version4k.leetcode

import com.only4.algorithm.version4k.extra.TreeNode

fun diameterOfBinaryTree(root: TreeNode?): Int {
    if (root == null) return 0

    var maxDiameter = 0

    fun calculateDepth(node: TreeNode?): Int {
        if (node == null) return 0

        val leftDepth = calculateDepth(node.left)
        val rightDepth = calculateDepth(node.right)

        // 更新最大直径
        maxDiameter = maxOf(maxDiameter, leftDepth + rightDepth)

        // 返回当前节点的最大深度
        return maxOf(leftDepth, rightDepth) + 1
    }

    calculateDepth(root)
    return maxDiameter
}
